### ISSUE TITLE
[IOS] [Bug/Question?] Missing option to disable annotation type freetext callout

### ISSUE THREAD

**Issue by matchwerk-dev** [PSPDFKit/react-native#305]
# Details

On Android there is an option to disable the 'freetext callout' annotation type. On IOS removing 'freetext' from the array of editable annotation types causes the removal of both, freetext (with a reference to a specific position) and the basic text field.

How can we only disable this annotation type:
<img width="244" alt="Screenshot 2019-09-30 at 16 12 16" src="https://user-images.githubusercontent.com/55987555/65886778-3c25ed80-e39d-11e9-8a40-9b1997913e6c.png">
?

# Acceptance Criteria

- [ ] [IOS] There is a possibility to remove the freetext callout annotation type without removing the text annotation itself.

**Comment by leoek** [comment]
@radazzouz This is related to #276 and #278. (Because of them it is really important to use that we can disable this type of annotations)

**Comment by radazzouz** [comment]
@matchwerk-dev @leoek,

This is quite an advanced customization use case. I am sorry that there's no easy way to achieve this using the React Native Javascript API. Exposing the change to the main repo is also a bit complex. So I created an internal feature request to revisit the native API in order to make it easier to expose it to React Native and other hybrid wrappers. I also added your votes to the feature request.

That being said, hiding the callout button from the toolbar can be achieved using native (Objective-C) API. I've pushed the changes to https://github.com/PSPDFKit/react-native/commit/4fa445a1017134b6e89bbba0686956fbf9034955 on the [customize-the-toolbar-in-native-code branch](https://github.com/PSPDFKit/react-native/tree/customize-the-toolbar-in-native-code). This is the case where it would be a lot easier to fork the repo and apply the changes in there. We wrote about it on our [blog here](https://pspdfkit.com/blog/2018/how-to-extend-react-native-api/#customizing-the-ui-using-native-code).

I am also attaching a modified version of [`RCTPSPDFKitView.m` here](https://github.com/PSPDFKit/react-native/files/3672383/RCTPSPDFKitView.m.txt) so you can easily integrate it into your project/fork.

Let me know if this helps you with your use case.

**Comment by leoek** [comment]
@radazzouz thanks for the fast response!

Just to confirm: We only want to disable another annotation type (freetext_callout). This should be pretty easy through the `editableAnnotationTypes Array in the configuration, or am I missing something?

(This worked pretty good for us with the other types of annotations)

**Comment by steipete** [comment]
I know this is a bit confusing because PDF:

Callouts are a spezialization of free text, and the `editableAnnotationTypes` only has a per-annotation-type granularity. So if you disable free text, you also disable the specializations; but if you only want to disable the callout variant, then you need to customize this on the toolbar directly.

**Comment by leoek** [comment]
@steipete 

> I know this is a bit confusing because PDF

Shouldn't this pdf implementation detail be abstracted away by pspdfkit(-ios)? On Windows ans Android this works fine and the available annotations can be configured without dealing with the specialties of the pdf spec.

I guess this is another issue for the list with incompatibilities between the different platforms.

**Comment by steipete** [comment]
Both iOS and Android have the concept of `editableAnnotationTypes` that work based on the annotation type:

https://pspdfkit.com/api/android/reference/com/pspdfkit/configuration/PdfConfiguration.Builder.html#editableAnnotationTypes(java.util.List%3Ccom.pspdfkit.annotations.AnnotationType%3E)
https://pspdfkit.com/guides/ios/current/annotations/configuring-editablevisible-annotation-types/

This filter is based on type; callout on both can't be specifically excluded because it's a specialization of the freetext type.

There's a second way that filters not on the annotation type, but on the tool/variant/specialization:

Android:
https://pspdfkit.com/api/android/reference/com/pspdfkit/configuration/PdfConfiguration.Builder.html#enabledAnnotationTools(java.util.List%3Ccom.pspdfkit.ui.special_mode.controller.AnnotationTool%3E)

iOS:
https://pspdfkit.com/guides/ios/current/customizing-the-interface/customize-the-annotation-toolbar/#annotation-buttons

We'll follow up with sample code how to hide callout from the annotation toolbar, and if that requires platform-specific API, we'll work on unifying that.

**Comment by radazzouz** [comment]
@leoek,

Have you tried using the sample code from https://github.com/PSPDFKit/react-native/issues/305#issuecomment-536655932 in your project?

**Comment by leoek** [comment]
@radazzouz 

Not yet. We are trying to minimize native code changes to be able to merge upstream changes without to much overhead. I will try that though.

@steipete 

> We'll follow up with sample code how to hide callout from the annotation toolbar, and if that requires platform-specific API, we'll work on unifying that.

That would be great. Currently we only use the `editableAnnotationTypes` configuration on IOS. On Windows it is not yet available, therefore we use `setToolbarItems()` there. On Android we are using the `menuItemGrouping` prop to configure which annotation options are shown to the user. Is the `editableAnnotationTypes` configuration usable thorugh the react-native wrapper on android? (If I remember correctly we tried that before we switched to the `menuItemGrouping` prop)

**Comment by irgendeinich** [comment]
@leoek 
You are correct `editableAnnotationTypes` is not exposed on Android currently. Is something preventing you from using `menuItemGrouping` on iOS as well?

Our current plan to remedy this situation is to make sure that you can remove the callout option by using the `menuItemGrouping` prop on iOS.

**Comment by leoek** [comment]
@irgendeinich 
> Is something preventing you from using menuItemGrouping on iOS as well?

Nothing that I know of, I did not realize that we could use that on iOS as well. Does the `menuItemGrouping` prop on iOS accept the same types as `menuItemGrouping` on android?

We are using these on android: 
```
allAndroidToolbarItems: [
      "markup",
      "writing",
      "highlight",
      "squiggly",
      "strikeout",
      "underline",
      "freetext",
      "freetext_callout",
      "signature",
      "ink",
      "note",
      "drawing",
      "multimedia",
      "image",
      "camera",
      "stamp",
      "line",
      "square",
      "circle",
      "polygon",
      "polyline",
      "erase",
      "redaction",
      "picker"
    ]
```

**Comment by irgendeinich** [comment]
You can see the supported types [here](https://github.com/PSPDFKit/react-native/blob/77dd26da6a78a6830918a2cba5d5adff9d18e682/ios/RCTPSPDFKit/Converters/RCTConvert%2BPSPDFAnnotationToolbarConfiguration.m#L39).


It looks like a couple are missing on iOS that exist on Android (`freetext_callout` immediately jumps out, also the groups like `multimedia`, `drawing`, etc.). So with this array you are enabling all options right? If I understood the ticket correctly you were trying to hide the callout option though so that not existing on iOS should be fine?

My suggestion would be you try using the `menuItemGrouping` on iOS as well and if the items you provide don't work like they should I'll prepare a PR that fixes up the converter accordingly.

**Comment by leoek** [comment]
@irgendeinich 

> So with this array you are enabling all options right?

Yes, that array would enable everything. (I just wanted to confirm the types, as they are capitalized for the `editableAnnotationTypes` configuration on iOS) To put this into a context, we build the configurations metnioned here https://github.com/PSPDFKit/react-native/issues/305#issuecomment-542230960 from a config object which looks like this (which is also the reason why this #203 would be a great improvement, as we could get rid of all these strings...): 
```
allToolbarItems: [
      "sidebar-thumbnails",
      "sidebar-document-outline",
      "sidebar-annotations",
      "sidebar-bookmarks",
      "pager",
      "layout-config",
      "pan",
      "zoom-out",
      "zoom-in",
      "zoom-mode",
      "spacer",
      "annotate",
      "ink",
      "text-highlighter",
      "highlighter",
      "ink-signature",
      "image",
      "stamp",
      "note",
      "text",
      "line",
      "arrow",
      "rectangle",
      "ellipse",
      "polygon",
      "polyline",
      "print",
      "search"
    ],
    toolbarItemsWhiteList: {
      [BINARYTYPE.<a noneditable document type>]: [
        "sidebar-thumbnails",
        "sidebar-document-outline",
        "pager",
        "layout-config",
        "pan",
        "zoom-out",
        "zoom-in",
        "zoom-mode",
        "search",
        "note"
      ],
      [BINARYTYPE.<a editable document type>]: [
        "sidebar-thumbnails",
        "sidebar-document-outline",
        "pager",
        "layout-config",
        "pan",
        "zoom-out",
        "zoom-in",
        "zoom-mode",
        "search"
      ]
    },
    toolbarItemsBlackList: {
      [BINARYTYPE.<another editable document type>]: [
        "print",
        "image",
        "sidebar-bookmarks",
        "ink-signature",
        "stamp",
        "line",
        "arrow",
        "rectangle",
        "ellipse",
        "polygon",
        "polyline"
      ]
    },
    allEditableAnnotationTypes: [
      "Link",
      "Highlight",
      "Underline",
      "Squiggly",
      "StrikeOut",
      "Text",
      "Caret",
      "FreeText",
      "Ink",
      "Square",
      "Circle",
      "Line",
      "Signature",
      "Stamp",
      "Eraser",
      "Image",
      "Widget",
      "FileAttachment",
      "Sound",
      "Polygon",
      "PolyLine",
      "RichMedia",
      "Screen",
      "Popup",
      "Watermark",
      "TrapNet",
      "3D",
      "Redact"
    ],
    editableAnnotationTypesWhiteList: {
      [BINARYTYPE.<a noneditable document type>]: [],
      [BINARYTYPE.<an editable document type>]: [
       "Text"
      ]
    },
    editableAnnotationTypesBlackList: {
      [BINARYTYPE.<another editable document type>]: [
        "Link",
        "Caret",
        "Square",
        "Circle",
        "Line",
        "Signature",
        "Stamp",
        "Image",
        "Widget",
        "FileAttachment",
        "Sound",
        "Polygon",
        "PolyLine",
        "RichMedia",
        "Screen",
        "Popup",
        "Watermark",
        "TrapNet",
        "3D",
        "Redact"
      ]
    },
    allAndroidToolbarItems: [
      "markup",
      "writing",
      "highlight",
      "squiggly",
      "strikeout",
      "underline",
      "freetext",
      "freetext_callout",
      "signature",
      "ink",
      "note",
      "drawing",
      "multimedia",
      "image",
      "camera",
      "stamp",
      "line",
      "square",
      "circle",
      "polygon",
      "polyline",
      "erase",
      "redaction",
      "picker"
    ],
    androidToolbarItemsWhiteList: {
      [BINARYTYPE.<a noneditable document type>]: [],
      [BINARYTYPE.<an editable document type>]: [
       "note"
      ]
    },
    androidToolbarItemsBlackList: {
      [BINARYTYPE.<another editable document type>]: [
        "freetext_callout",
        "signature",
        "multimedia",
        "image",
        "camera",
        "stamp",
        "line",
        "square",
        "circle",
        "polygon",
        "polyline",
        "redaction",
        "picker"
      ]
    }
```

**Comment by Elmatoss** [comment]
@irgendeinich 

I have tried using the `menuItemGrouping` on iOS and encountered a few problems:

- The following toolbaritems are not available: `markup`, `writing`, `freetext_callout`, `multimedia`, `camera`, `picker`
- The last 4 don't matter to us, but we need `markup` and `writing`, so adding those would be awesome
- The spelling of all the strings are the same as on android (all lowercase), which we *highly* appreciate, BUT on android the rubber/eraser item is called `erase` and on iOS `eraser`. That is pretty weird concerning everything else is exactly the same

So in order for the `menuItemGrouping` to work for us we would need: 
- The toolbarItems `markup` and `writing`
- Consistent spelling (erase OR eraser)
